### ISSUE TITLE
Runtime error in exhaustive switch

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#7215]
The following code doesn't work well in HL because the switch is not exhaustive at runtime:

```haxe
enum abstract A(String) {
	var A;
	var B;
}

class Test {
	static function foo( e : A ) return { v : switch( e ) { case A: 55; case B: 66; } };
	static function main() {
		trace(foo(null));
		trace(foo(cast "C")); // if loaded from JSON for example
	}
}
```

I wonder if some other platforms are concerned - in which case this should be fixed in Haxe - I expect both results to be `{ v : 0 }`. BTW it is not reproducible with simply "return switch"

**Comment by RealyUniqueName** [comment]
That's what you get for using Dynamic or unsafe cast :)
I think this should be either unspecified behaviour or throw an error on `switch` like "Invalid enum value" to prevent propagating subtle runtime errors all over the application. But that would make any function with `switch` not pure.

Similar problem exists with switching on a normal enum. It's up to developer to make sure it's not `null` before switching.

**Comment by Simn** [comment]
Maybe we could tag the enum abstract as `@:infinite` or something and require a catch all then. That's also how OCaml extensible variants work.

**Comment by RealyUniqueName** [comment]
Nice idea to have infinite enums. That would be great e.g. for http headers, mime types etc.

But as for the case described in the first post I think it's up to developer to validate json data or any other data obtained through Dynamic or unsafe cast.

**Comment by ncannasse** [comment]
The problem is not really about type safety here, but about HL returning values such as "1489130616" because of the way it compiles the switch. That can produces hard-crash when returning values instead of integers (random adress ptr)

**Comment by Simn** [comment]
I think HL should fix that. Why would we silently default to 0 here? A crash is arguably better.

**Comment by ncannasse** [comment]
HL should never crash. So when such a switch occurs there's two possible
behaviors: either throw an exception "Unhandled value X" or return the
default value for the given return type (0/null/false).
Anyway we need to have something consistant across platforms, so it would
be better handling at Haxe level, not at HL one.

**Comment by Simn** [comment]
So what you're saying is that we always have to generate a `default: throw "Unmatched pattern"`.

**Comment by ncannasse** [comment]
In that specific case, yes.

**Comment by Simn** [comment]
What's "specific" about this case? How about a normal enum switch that doesn't account for the enum value being `null`?

**Comment by ncannasse** [comment]
What I meant by "specific" is having switches on enum abstract as an
R-value (returning a value).
For normal enum we usually switch on the index so it creates a null access
if we switch on null.

**Comment by kevinresol** [comment]
> it creates a null access if we switch on null.

Which is actually a very hard to understand error to newcomers

**Comment by Simn** [comment]
Let's see:

cpp, hl: random values
neko, js, python, php, eval: null
flash, java, cs: 0
jvm: first case

This tells me that we should just fix it in our generators.

**Comment by ncannasse** [comment]
I'm fine with whatever fix.

**Comment by Simn** [comment]
Looking at this again, it's not something the generators could fix because the switch has been transformed by the time it reaches them:

```
	static function foo(e:A) {
		var tmp;
		switch ((@:exhaustive e)) {
			case "A": tmp = 55;
			case "B": tmp = 66;
		};
		return {v : tmp};
	}
```

I guess this means the pattern matcher has to generate a `default: throw` after all.

**Comment by Simn** [comment]
And then the question is if we should do this only for cpp/hl/jvm or for all targets. I don't like inconsistencies and silent null values, so I'm thinking the latter.

**Comment by ncannasse** [comment]
Yes, if it's an exception it should be consistent across targets.
